### PR TITLE
Add `safe_execution_block_hash` property to FCR tests

### PR DESCRIPTION
**Pull request by mkalinin**
The main purpose of this PR is to add `safe_execution_block_hash` check to the FCR tests which if implemented by client implementations allow to check whether the right execution block_hash is deemed safe by the CL.

This is especially important for Gloas, where `get_safe_execution_block_hash` function is modified and returns a `block_hash` of the parent payload of the confirmed block.

This new check requires `safe_execution_block_hash` to get updated each time the confirmation chain advances which was not the case for Gloas as the `block_hash` was set to a constant value in test helpers. This PR attempts to fix this by building a payload at the same time as the beacon block is built and incorporating the parent payload into this built process. The genesis case is handled in the following way: a payload produced by `get_sample_genesis_execution_payload` is treated as the payload of the genesis’ parent (which is a full block) and let’s the genesis block to be an empty block — i guess this is the convention around handling genesis block in Gloas and after.

I think the part of the test generation related to execution payload requires a redesign as the Gloas affects it significantly and atm it looks messy. But this redesign is out of the scope of this PR.

This PR also introduces a couple of fields that are new to Gloas to the payload `block_hash` computation and another one FCR test that potentially covers the edge case that wasn’t covered before.

### CODE REVIEW

File: tests/core/pyspec/eth_consensus_specs/test/helpers/block.py [hunk 1]
```diff
@@ -129,6 +130,62 @@ def build_empty_block(spec, state, slot=None, proposer_index=None):
...
+    apply_randao_reveal(spec, state, block, proposer_index)
+
+    if is_post_altair(spec):
+        block.body.sync_aggregate.sync_committee_signature = spec.G2_POINT_AT_INFINITY
+
+    if not is_post_bellatrix(spec):
+        return block, None
+
+    if is_post_electra(spec) and execution_requests is None:
+        execution_requests = spec.ExecutionRequests()
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/block.py:169]: Do you think it would be valuable to add an `else: assert execution_requests is None` after this? To ensure we don't accidentally provide execution requests pre-electra?

Comment (jtraglia) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/block.py:169]: To be clear, I'd probably rewrite it as:

```python
if is_post_electra(spec):
    if execution_requests is None:
        execution_requests = spec.ExecutionRequests()
else:
    assert execution_requests is None
```

File: tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py [hunk 2]
```diff
@@ -89,68 +91,71 @@ def compute_requests_hash(block_requests):
...
         # parent_beacon_root
         execution_payload_header_rlp.append((Binary(32, 32), parent_beacon_block_root))
     if is_post_electra(spec):
         # requests_hash
         execution_payload_header_rlp.append((Binary(32, 32), requests_hash))
+    if is_post_gloas(spec):
+        # block access list
+        execution_payload_header_rlp.append((Binary(32, 32), block_access_list_hash))
+        # slot number
+        execution_payload_header_rlp.append((big_endian_int, payload.slot_number))
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py:158]: Nice catch, thanks 🙏

File: tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py [hunk 3]
```diff
@@ -310,38 +320,56 @@ def build_empty_post_gloas_execution_payload_bid(spec, state):
     )
 
 
-def build_empty_signed_execution_payload_bid(spec, state):
-    if not is_post_gloas(spec):
-        return
-    message = build_empty_post_gloas_execution_payload_bid(spec, state)
+def sign_execution_payload_bid(spec, state, bid):
+    assert is_post_gloas(spec)
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py:324]: Just wanted to comment in support of using `asserts` here instead of simply returning, this is a much better pattern.

File: tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py [hunk 4]
```diff
@@ -310,38 +320,56 @@ def build_empty_post_gloas_execution_payload_bid(spec, state):
...
-        message=message,
+        message=bid,
         signature=signature,
     )
 
 
-def build_empty_execution_payload(spec, state, randao_mix=None):
+def build_empty_signed_execution_payload_bid(spec, state):
+    if not is_post_gloas(spec):
+        return
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/execution_payload.py]: Hmm should this be `assert is_post_gloas(spec)` too? I feel like we shouldn't call this function pre-Gloas.

File: tests/core/pyspec/eth_consensus_specs/test/helpers/fast_confirmation.py [hunk 5]
```diff
@@ -187,6 +197,28 @@ def max_attestations(self):
         else:
             return self.spec.MAX_ATTESTATIONS
 
+    def get_parent_payload(self, parent_root):
+        if not is_post_bellatrix(self.spec):
+            return None
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/fast_confirmation.py]: Maybe this can be `assert is_post_bellatrix(self.spec)` too?

File: tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py [hunk 6]
```diff
@@ -262,11 +262,8 @@ def get_genesis_forkchoice_store_and_block(spec, genesis_state):
         # Match the genesis block body bid to what ``genesis.py`` set on the
         # state's committed bid; this keeps ``genesis_block`` consistent with
         # ``genesis_state.latest_block_header`` (body_root).
-        genesis_block.body.signed_execution_payload_bid.message = spec.ExecutionPayloadBid(
-            # The genesis bid's block hash is the empty hash
-            block_hash=spec.Hash32(),
-            parent_block_hash=spec.Hash32(genesis_state.latest_block_hash),
-            execution_requests_root=spec.hash_tree_root(spec.ExecutionRequests()),
+        genesis_block.body.signed_execution_payload_bid.message = (
+            genesis_state.latest_execution_payload_bid
```
Comment (jtraglia) [inline tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py:266]: Just to confirm, this re-write doesn't change the field values, correct? The behavior here was debated for quite a while and I do not wish to restart that 😄

File: tests/core/pyspec/eth_consensus_specs/test/helpers/block.py (hunk 1 above)
Comment (mkalinin) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/block.py:169]: will add this precondition to the top of the function body

File: tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py (hunk 6 above)
Comment (mkalinin) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py:266]: Please take a look at these lines:
https://github.com/ethereum/consensus-specs/pull/5449/changes#diff-40664692fff1984104005f7a05d7305c2e516c8b45e82faf809c8ac2a6ed1b22R178

I think what matters the most is `block_hash=spec.Hash32()`, it shouldn’t matter much which exact value will `parent_block_hash` accept. I think what matters is that the genesis block is an _empty_ block

File: tests/core/pyspec/eth_consensus_specs/test/helpers/block.py (hunk 1 above)
Comment (jtraglia) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/block.py:169]: That's even better, thanks 🙏

File: tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py (hunk 6 above)
Comment (jtraglia) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py:266]: I see thanks for the link, yeah `parent_block_hash` will be different. Hmm I think you're right, but let me confirm with @nflaig too.

Comment (nflaig) [reply tests/core/pyspec/eth_consensus_specs/test/helpers/fork_choice.py:266]: let me try to summarize this

current spec
- `state.latest_block_hash = eth1_block_hash`
- `state.latest_execution_payload_bid.parent_block_hash = eth1_block_hash`
- `state.latest_execution_payload_bid.block_hash = Hash32()`

after this PR
```python
genesis_payload = get_sample_genesis_execution_payload(...)
G = genesis_payload.block_hash
```
- `state.latest_block_hash = G`
- `state.latest_execution_payload_bid.parent_block_hash = G`
- `state.latest_execution_payload_bid.block_hash = Hash32()`

so the genesis block will still be EMPTY, what changes here is only the `parent_block_hash` which no longer references the `eth1_block_hash` in the spec tests, is that correct?

worth pointing out that on a real genesis state (eg. on kurtosis with actual EL client), the `parent_block_hash` needs to reference the actual EL genesis block hash as the first produced payload needs to be build on top of that

**Review by nflaig — APPROVED** [review]
LGTM, based on https://github.com/ethereum/consensus-specs/pull/5449#discussion_r3630047613, if my understanding is correct, the changes should not cause any issues for clients

**Review by jtraglia — APPROVED** [review]
